lib/inspec/config.rb

```diff
@@ -0,0 +1,286 @@
+# Represents InSpec configuration.  Merges defaults, config file options,
+# and CLI arguments.
+
+require 'pp'
+
+module Inspec
+  class Config
+    # These are options that apply to any transport
+    GENERIC_CREDENTIALS = %w{
+      backend
+      sudo
+      sudo_password
+      sudo_command
+      sudo_options
+      shell
+      shell_options
+      shell_command
+    }.freeze
+
+    extend Forwardable
+
+    # Many parts of InSpec expect to treat the Config as a Hash
+    def_delegators :@final_options, :each, :delete, :[], :[]=, :key?
+    attr_reader :final_options
+
+    def initialize(cli_opts = {}, cfg_io = nil, command_name = nil)
+      @command_name = command_name || (ARGV.empty? ? nil : ARGV[0].to_sym)
+      @defaults = Defaults.for_command(@command_name)
+
+      @cli_opts = cli_opts.dup
+      cfg_io = resolve_cfg_io(@cli_opts, cfg_io)
+      @cfg_file_contents = read_cfg_file_io(cfg_io)
+
+      @merged_options = merge_options
+      @final_options = finalize_options
+    end
+
+    def diagnose
+      return unless self[:diagnose]
+      puts "InSpec version: #{Inspec::VERSION}"
+      puts "Train version: #{Train::VERSION}"
+      puts 'Command line configuration:'
+      pp @cli_opts
+      puts 'JSON configuration file:'
+      pp @cfg_file_contents
+      puts 'Merged configuration:'
+      pp @merged_options
+      puts
+    end
+
+    #-----------------------------------------------------------------------#
+    #                         Reading Config Files
+    #-----------------------------------------------------------------------#
+
+    # Regardless of our situation, end up with a readable IO object
+    def resolve_cfg_io(cli_opts, cfg_io)
+      unless cfg_io
+        # TODO: deprecate --json-config
+        path = cli_opts[:config] || cli_opts[:json_config]
+        if path == '-'
+          Inspec::Log.warn 'Reading JSON config from standard input' if STDIN.tty?
+          path = STDIN
+        elsif path.nil?
+          default_path = File.join(Inspec.config_dir, 'config.json')
+          path = default_path if File.exist?(default_path)
+        elsif !File.exist?(path)
+          raise ArgumentError, "Could not read configuration file at #{path}"
+        end
+        cfg_io = File.open(path) if path
+        cfg_io ||= StringIO.new('{ "version": "1.1" }')
+      end
+      cfg_io
+    end
+
+    def read_cfg_file_io(cfg_io)
+      contents = cfg_io.read
+      begin
+        @cfg_file_contents = JSON.parse(contents)
+        validate_config_file_contents!
+      rescue JSON::ParserError => e
+        raise Inspec::ConfigError::MalformedJson, "Failed to load JSON configuration: #{e}\nConfig was: #{contents}"
+      end
+      @cfg_file_contents
+    end
+
+    def file_version
+      @cfg_file_contents['version'] || :legacy
+    end
+
+    def legacy_file?
+      file_version == :legacy
+    end
+
+    def config_file_cli_options
+      if legacy_file?
```

Comment:
I don't see where `file_version` or `legacy_file` are used anywhere else. 
Any reason we shouldn't remove those methods and change this to?:

```
# Assume legacy file if no version is specified
if !@cfg_file_contents['version']
```

Reply:
A big point of this PR series is to formalize the config file format, and get it into a shape where can version the format, and validate it.

Having the Config object be able to report on the detected version is thus key to the intent of the PR.